src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnUserAccountService.java

```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Transactional(readOnly = true)
+    public WebAuthnUserAccount findByProviderAndSubject(String provider, String subject) throws NoSuchUserException {
+        WebAuthnUserAccount account = accountRepository.findByProviderAndSubject(provider, subject);
+        if (account == null) {
+            throw new NoSuchUserException();
+        }
+        return accountRepository.detach(account);
+    }
+
+    @Transactional(readOnly = true)
+    public WebAuthnUserAccount findByCredentialId(String credentialId) {
+        WebAuthnCredential cred = credentialRepository.findByCredentialId(credentialId);
+        if (cred == null) {
+            return null;
+        }
+        WebAuthnUserAccount userAccount = accountRepository.getOne(cred.getUserHandle());
+        return accountRepository.detach(userAccount);
+    }
+
+    /*
+     * CRUD
+     */
+    public WebAuthnUserAccount addAccount(
+            WebAuthnUserAccount reg) throws RegistrationException {
+
+        try {
+            // we explode model
+            WebAuthnUserAccount account = new WebAuthnUserAccount();
+
+            account.setEmailAddress(reg.getEmailAddress());
+            account.setProvider(reg.getProvider());
+            account.setRealm(reg.getRealm());
+            account.setUserHandle(reg.getUserHandle());
+            account.setUsername(reg.getUsername());
+
+            account = accountRepository.saveAndFlush(account);
+            account = accountRepository.detach(account);
+
+            return account;
+        } catch (Exception e) {
+            throw new RegistrationException(e.getMessage());
+        }
+    }
+
+    public WebAuthnUserAccount updateAccount(
+            String userHandle,
+            WebAuthnUserAccount reg) throws NoSuchUserException, RegistrationException {
+        if (reg == null) {
+            throw new NoSuchUserException();
+        }
+
+        WebAuthnUserAccount account = accountRepository.findByUserHandle(userHandle);
+        if (account == null) {
+            throw new NoSuchUserException();
+        }
+
+        try {
+            // we explode model and update every field
+            account.setEmailAddress(reg.getEmailAddress());
+            account.setProvider(reg.getProvider());
+            account.setRealm(reg.getRealm());
+            account.setUserHandle(reg.getUserHandle());
+            account.setUsername(reg.getUsername());
+
+            account = accountRepository.saveAndFlush(account);
+            account = accountRepository.detach(account);
+
+            return account;
+        } catch (Exception e) {
+            throw new RegistrationException(e.getMessage());
+        }
+    }
+
+    public void deleteAccount(String provider, String subject) {
+        final WebAuthnUserAccount account = accountRepository.findByProviderAndSubject(provider, subject);
+        for (final WebAuthnCredential c : credentialRepository.findByUserHandle(account.getUserHandle())) {
```

Comment:
this construction is incorrect: besides being expensive could lead to side effects while you remove elements from repo inside the loop.
You should first fetch all the credentials by calling once findByUserHandle, and then either loop over the results or simply call deleteAll/deleteInBatch
```
      List<WebAuthnCredential> credentials = credentialRepository.findByUserHandle(account.getUserHandle());
        credentialRepository.deleteInBatch(credentials);
```

Reply:
Fixed in 89f3a06fe0e2a665bc7a401d43a42720e7b1f80f